File: solution/1400-1499/1416.Restore The Array/Solution.java
class Solution {
    private Integer[] f;
    private String s;
    private int n;
    private int k;

    public int numberOfArrays(String s, int k) {
        n = s.length();
        this.s = s;
        this.k = k;
        f = new Integer[n];
        return dfs(0);
    }
    
    private int dfs(int i) {
        if (i >= n) {
            return 1;
        }
        if (s.charAt(i) == '0') {
            return 0;
        }
        if (f[i] != null) {
            return f[i];
        }
        int ans = 0;
        long x = 0;
        final int mod = (int) 1e9 + 7;
        for (int j = i; j < n; ++j) {
            x = x * 10 + s.charAt(j) - '0';
            if (x > k) {
                break;
            }
            ans = (ans + dfs(j + 1)) % mod;
        }
        return f[i] = ans;
    }
}

File: solution/1400-1499/1416.Restore The Array/Solution.ts
function numberOfArrays(s: string, k: number): number {
    const n = s.length;
    const f: number[] = new Array(n + 1).fill(-1);
    const dfs = (i: number): number => {
        if (i >= n) {
            return 1;
        }
        if (s[i] === '0') {
            return 0;
        }
        if (f[i] !== -1) {
            return f[i];
        }
        let x = 0;
        let ans = 0;
        for (let j = i; j < n; ++j) {
            x = x * 10 + s[j].charCodeAt(0) - '0'.charCodeAt(0);
            if (x > k) {
                break;
            }
            ans += dfs(j + 1);
            ans %= 1000000007;
        }
        return (f[i] = ans);
    };
    return dfs(0);
}
